build-logic/buildquality/src/main/kotlin/gradlebuild/testcleanup/TestFilesCleanupService.kt

```diff
@@ -178,67 +95,93 @@ abstract class TestFilesCleanupService @Inject constructor(
-                        }
+                    addExecutedTaskPath(taskPath)
+                    if (containsFailedTest(taskPath)) {
+                        addFailedTaskPath(taskPath)
                     }
                 }
                 is TaskFailureResult -> {
-                    executedTaskPaths.add(taskPath)
-                    failedTaskPaths.add(taskPath)
+                    addExecutedTaskPath(taskPath)
+                    addFailedTaskPath(taskPath)
                 }
                 else -> {
                 }
             }
         }
     }
 
+    private
+    fun containsFailedTest(taskPath: String): Boolean {
+        return testPathToBinaryResultsDirs[taskPath]?.let { containsFailedTest(it) } == true
+    }
+
+    private
+    fun addExecutedTaskPath(taskPath: String) {
+        projectPathToExecutedTaskPaths.computeIfAbsent(relevantTaskPathToProjectPath.getValue(taskPath)) { mutableListOf() }.add(taskPath)
+    }
+
+    private
+    fun addFailedTaskPath(taskPath: String) {
+        projectPathToFailedTaskPaths.computeIfAbsent(relevantTaskPathToProjectPath.getValue(taskPath)) { mutableListOf() }.add(taskPath)
+    }
+
+    private
+    fun getFailedTaskPaths(projectPath: String) = projectPathToFailedTaskPaths.getOrDefault(projectPath, emptyList())
+
+    private
+    fun getExecutedTaskPaths(projectPath: String) = projectPathToExecutedTaskPaths.getOrDefault(projectPath, emptyList())
+
     override fun close() {
-        val tmpTestFiles = tmpTestFiles(buildDir)
-        prepareReportsForCiPublishing(
-            if (tmpTestFiles.isEmpty()) failedTaskPaths else executedTaskPaths,
-            executedTaskPaths,
-            tmpTestFiles.keys
-        )
-        cleanUp(tmpTestFiles.keys)
+        val projectPathToLeftoverFiles = mutableMapOf<String, LeftoverFiles>()
+        // First run: collect and archieve leftover files
+        parameters.projectExtensions.get().forEach { (projectPath: String, projectExtension: TestFileCleanUpExtension) ->
+            val tmpTestFiles = projectExtension.tmpTestFiles()
+            if (tmpTestFiles.isEmpty()) {
+                return@forEach
+            }
+
+            projectExtension.prepareReportsForCiPublishing(
+                if (tmpTestFiles.isEmpty()) getFailedTaskPaths(projectPath) else getExecutedTaskPaths(projectPath),
+                getExecutedTaskPaths(projectPath),
+                tmpTestFiles.keys
+            )
+            cleanUp(tmpTestFiles.keys)
+            projectPathToLeftoverFiles[projectPath] = tmpTestFiles
+        }
+
+        // Second run: verify and throw exceptions
         if (!cleanupRunnerStep) {
-            verifyTestFilesCleanup(failedTaskPaths, tmpTestFiles)
+            parameters.projectExtensions.get()
+                .filter { projectPathToLeftoverFiles.containsKey(it.key) }
+                .forEach { (projectPath: String, projectExtension: TestFileCleanUpExtension) ->
```

Comment:
I wonder if we could get all reports here, not only the first one.

Reply:
Now we collect all exceptions and throw a `MultiCauseException` at the end.